src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -84,5 +224,187 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+        /// <returns>A list of dependencies, indexed by framework/RID</returns>
+        /// <remarks>Assets file reading occurs in a background thread</remarks>
+        /// <seealso cref="GetAssetsFilePathAsync"/>
+        private async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(string assetsFilePath, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
+        /// </summary>
+        /// <param name="top">Top, Direct dependency</param>
+        /// <param name="current">Current package/node to visit</param>
+        /// <param name="graph">Package dependency graph, from assets file</param>
+        /// <param name="visited">Dictionary to remember visited nodes</param>
+        /// <param name="fxRidEntry">Framework/Runtime-ID associated with current <paramref name="graph"/></param>
+        private void MarkTransitiveOrigin(PackageReference top, PackageReference current, LockFileTarget graph, HashSet<PackageIdentity> visited, FrameworkRuntimePair fxRidEntry, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = default;
+
+            // Find first target node that matches current
+            foreach (LockFileTargetLibrary lib in graph.Libraries)
+            {
+                if (lib.Type == LibraryType.Package.Value
+                    && string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase))
+                {
+                    node = lib;
+                    break;
+                }
+            }
+
+            /* 
+             * && ((current.HasAllowedVersions && current.AllowedVersions.Satisfies(lib.Version)) ||
+                        (current.PackageIdentity.HasVersion && current.PackageIdentity.Version.Equals(lib.Version)))
+             */
+
+            if (node != default)
+            {
+                visited.Add(current.PackageIdentity); // visited
+
+                // Lookup Transitive Origins Cache
+                TransitiveEntry cachedEntry;
+                if (!TransitiveOriginsCache.TryGetValue(current.PackageIdentity.Id, out cachedEntry))
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
+                    {
+                        [fxRidEntry] = new List<PackageReference>()
+                    };
+                }
+
+                if (!cachedEntry.ContainsKey(fxRidEntry))
+                {
+                    cachedEntry[fxRidEntry] = new List<PackageReference>();
+                }
+
+                if (!cachedEntry[fxRidEntry].Contains(top))
```

Comment:
Please note here we're doing O(n) search since cachedEntry[fxRidEntry] is List, maybe ok for now, you don't have to change.

Reply:
Added comment: `// Dictionary value is a List. If perf. is bad, change to Dict.`

Reply:
Currently it's `list` so `HashSet` is enough, no need Dictionary.